utilities/scale-json/src/lib.rs

```diff
@@ -0,0 +1,391 @@
+use scale_decode::{
+	visitor::{decode_with_visitor, DecodeAsTypeResult, DecodeError},
+	IntoVisitor, TypeResolver,
+};
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+
+/// Re-export useful types from transitive dependencies.
+pub mod ext {
+	pub use scale_decode::DecodeAsType;
+	pub use serde_json::Value as JsonValue;
+}
+pub use scale_decode::Error;
+
+mod test;
+
+// TODO:
+// - Consider using the (newly added) composite name() or path() to improve decoding of (for
+//   example) account_ids. Or maybe add it as an additional __TYPE_HINT__ field in the JSON output.
+
+/// A wrapper type around a `serde_json::Value` that allows decoding scale encoded data into JSON.
+///
+/// Byte arrays are decoded as a single hex string (rather than an array of numbers).
+/// Numbers up to 32 bits wide are decoded as JSON Numbers, anything 64 bits wide or above is
+/// decoded as hex-encoded JSON Strings.
+///
+/// A heuristic is applied to arrays and sequences to determine if they should be interpreted as
+/// byte arrays: If the sequence or array is not empty, *and* if each contained item decodes to a
+/// json Number value, *and* each item was encoded as a single byte, then the sequence or array is
+/// interpreted as a sequence or array of bytes.
+///
+/// # Example
+///
+/// ```
+/// use scale_json::{
+///     ScaleDecodedToJson,
+///     ext::{DecodeAsType, JsonValue},
+/// };
+/// use scale_info::{PortableRegistry, TypeInfo};
+/// use codec::{Encode, Decode};
+///
+/// fn make_type_resolver<T: TypeInfo + 'static>() -> (u32, PortableRegistry) {
+///     let m = scale_info::MetaType::new::<T>();
+///     let mut registry = scale_info::Registry::new();
+///     let type_id = registry.register_type(&m).id;
+///     (type_id, PortableRegistry::from(registry))
+/// }
+///
+/// let (type_id, registry) = make_type_resolver::<Vec<u32>>();
+///
+/// let scale_encoded = vec![0u32, 1, 2, 3].encode();
+///
+/// let decoded_json: JsonValue = <ScaleDecodedToJson as DecodeAsType>::decode_as_type(
+///    &mut &scale_encoded[..],
+///    type_id,
+///    &registry,
+/// ).unwrap().into();
+///
+/// assert!(decoded_json.is_array());
+/// assert_eq!(decoded_json[2], JsonValue::Number(2.into()));
+/// ```
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ScaleDecodedToJson(serde_json::Value);
+
+impl From<serde_json::Value> for ScaleDecodedToJson {
+	fn from(value: serde_json::Value) -> Self {
+		Self(value)
+	}
+}
+impl From<ScaleDecodedToJson> for serde_json::Value {
+	fn from(value: ScaleDecodedToJson) -> Self {
+		value.0
+	}
+}
+impl AsRef<serde_json::Value> for ScaleDecodedToJson {
+	fn as_ref(&self) -> &serde_json::Value {
+		&self.0
+	}
+}
+
+// Implementing this is what gives us `DecodeAsType` for `ScaleDecodedToJson`.
```

Comment:
And i'm assuming this is why we also need the wrapper type, because we otherwise can't implement a foreign trait on a foreign type?